### PR TITLE
Implement event stream backed by streaming gRPC service.

### PR DESCRIPTION
**Pull request by jmalloc**
@danilvpetrov This PR needs tests yet, but I thought I'd get you in particular to take a look because there's some very "fancy" context shuffling going on just to get the gRPC streaming interfaces to look like regular Go code and I know you've dealt with this before. 

I want to make sure the inline comments are really helpful in understanding it even if they are verbose (assuming the reader already understands `context.Context`).

### CODE REVIEW

**Comment by jmalloc** [comment]
The stream interfaces in the `eventstream` package are very similar to those in [Aperture's `ordered` package](https://github.com/dogmatiq/aperture/blob/master/ordered/stream.go), but include full message meta-data, similar to Ax.

These interfaces are intended to provide an abstraction to event messages regardless of whether they come from the "local" persistence for "this" application, or across the network from another application.

File: api/stream.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+// Close stops the cursor.
+//
+// Any current or future calls to Next() return a non-nil error.
+func (c *cursor) Close() error {
+	c.cancel()
+	return nil
+}
+
+// consume waits receives new messages, unmarshals them and pipes them over the
```
Comment (danilvpetrov) [inline api/stream.go]:
```suggestion
// consume waits, receives new messages, unmarshals them, and pipes them over the
```

File: persistence/stream.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	// Any current or future calls to Next() return a non-nil error.
+	Close() error
+}
+
+// StreamMessage is a message on a stream.
+type StreamMessage struct {
+	// Offset is the offset of the message on the stream.
+	Offset uint64
+
+	// Envelope contains the message and its meta-data.
```
Comment (danilvpetrov) [inline persistence/stream.go]:
```suggestion
	// Envelope contains the message and its metadata.
```

Comment (jmalloc) [reply persistence/stream.go]: I've written this as `meta-data` everywhere else, so I'll leave that can of worms alone :)

File: api/stream.go (hunk 1 above)
Comment (jmalloc) [reply api/stream.go]: `waits` was an extra word... I've added your Oxford comma, though ;)

File: api/stream.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+	// consumeCtx lives for the lifetime of the stream returned by the gRPC
+	// Consume() operation. This is how we cancel the gRPC stream, as it has no
+	// Close() method.
+	//
+	// It does NOT use ctx as a parent, because ctx is intended only to span the
+	// lifetime of this call to Open().
+	consumeCtx, cancelConsume := context.WithCancel(context.Background())
+
+	// done is signalling channel used to indicate that the call to Consume()
```
Comment (jmalloc) [inline api/stream.go]:
```suggestion
	// done is a signalling channel used to indicate that the call to Consume()
```

**Comment by jmalloc** [comment]
I will work this into a future PR.

**Pull request closed without merging**